DPL: introduce a O2DatabasePDG with all the customizations

ktf opened the pull request:
This is to be used in place of the TDatabasePDG in the Analysis tasks.

jgrosseo commented:
This will bring down the memory I assume?

ktf commented:
In principle yes. The draw back is that it requires changes in O2Physics from:

```c++
Service<TDatabasePDG>
```

```c++
Service<o2::framework::O2DatabasePDG>
```

jgrosseo commented:
We could do this globally once we have validated it with 1 or 2 tasks.

@ddobrigk What do you think? This could be a topic for Monday I think.

ddobrigk commented:
Hi @ktf , @jgrosseo , yes, it would be great to touch base on this! Maybe you can prepare - in addition to your usual slides - a slide explaining what you did exactly in this PR? Thanks a lot!

ktf commented:
If you want to do that incrementally, we could leave in the:

```c++
    CommonAnalysisServices::addAnalysisService<T>(specs);
```

which will make things behave as before we migrate the tasks after validation and then we remove the above line which is what effectively adds the extra memory everywhere a service is instanciated.

jgrosseo commented:
Yes, makes sense. In this way we also do not have to force PRs... (or maybe I am missing something because fullCI passed actually)

ktf commented:
Full CI passes because we probably do not test any PDG related task in there. This is only for analysis. I will update it in a bit.

jgrosseo commented:
Ah so it would not work at run time. Now I got it...

ktf commented:
I would suggest we merge this, since the behaviour should be unchanged if you do not use the `Service<O2DatabasePDG>` . I tested on hyperloop and in both cases it add the particle 3124 (twice, BTW...), which is one of our customizations.

ktf commented:
We can then enable and validate the on demand customizations in a couple of tasks and finally remove the old customizations.

sawenzel commented on Framework/PhysicsSupport/include/Framework/O2DatabasePDG.h:
```diff
@@ -0,0 +1,26 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
```
This would lead to 2 header files of the same name `O2DatabasePDG.h` in O2. Maybe a better name could be found? (O2DatabasePDGPlugin ... or similar)?

ktf replied:
ok.

ktf committed:
DPL Analysis: load PDG database on demand

Old version still behaves as before, but it will go away as
soon as we move O2Physics to the on demand behavior.

ktf commented:
The unit test in #10538 actually shows a regression when using the old method (new method works fine). I am holding on this until it's fixed.

ktf commented:
PR updated. The unit test shows which both old and new version behave the same, including the customizations. Will now provide a PR for O2Physics while this gets tested.

ktf commented:
There were some more issues when running on hyperloop which the new update fixes.

ktf commented:
Merging since the o2 test seems to be stuck and this was tested on hyperloop as well. The fact that the new Service has exactly the same behaviour as the old one should be guaranteed by the unit test.

knopers8 commented:
@ktf It seems it broke all the builds in QC:
```
[13/1345] Building CXX object Framework/PhysicsSupport/CMakeFiles/O2lib-FrameworkPhysicsSupport.dir/src/Plugin.cxx.o
FAILED: Framework/PhysicsSupport/CMakeFiles/O2lib-FrameworkPhysicsSupport.dir/src/Plugin.cxx.o 
/sw/slc7_x86-64/GCC-Toolchain/v10.2.0-alice2-9/bin/c++ -DBOOST_ALL_NO_LIB -DBOOST_CONTAINER_DYN_LINK -DBOOST_ERROR_CODE_HEADER_ONLY -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_IOSTREAMS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_REGEX_DYN_LINK -DDPL_ENABLE_BACKTRACE -DENABLE_UPGRADES -DFAIRMQ_HAS_STD_FILESYSTEM=1 -DFMT_SHARED -DO2lib_FrameworkPhysicsSupport_EXPORTS -DZDC_FASTSIM_ONNX -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/Framework/PhysicsSupport/src -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/Framework/Core/include -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/DataFormats/simulation/include -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/Framework/Foundation/include -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/Framework/Foundation/3rdparty/include -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/Common/Constants/include -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/DataFormats/Headers/include -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/DataFormats/MemoryResources/include -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/GPU/Common -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/Framework/Logger/include -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/Utilities/PCG/include -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/DataFormats/Detectors/Common/include -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/GPU/Utils -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/Common/MathUtils/include -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/Utilities/rANS/include -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/Common/Utils/include -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/DataFormats/common/include -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/DataFormats/Parameters/include -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/Common/Types/include -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/CCDB/include -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/DataFormats/Calibration/include -I/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/DataFormats/Reconstruction/include -isystem /sw/slc7_x86-64/Configuration/v2.6.3-46/include -isystem /sw/slc7_x86-64/boost/v1.75.0-95/include -isystem /sw/slc7_x86-64/Monitoring/v3.15.1-4/include -isystem /sw/slc7_x86-64/curl/7.70.0-48/include -isystem /sw/slc7_x86-64/FairMQ/v1.4.55-18/include/fairmq -isystem /sw/slc7_x86-64/FairMQ/v1.4.55-18/include -isystem /sw/slc7_x86-64/FairLogger/v1.11.1-4/include -isystem /sw/slc7_x86-64/fmt/9.1.0-1/include -isystem /sw/slc7_x86-64/ROOT/v6-26-10-alice5-5/include -isystem /sw/slc7_x86-64/RapidJSON/v1.1.0-alice2-36/include -isystem /sw/slc7_x86-64/arrow/v9.0.0-alice1-24/include -isystem /sw/slc7_x86-64/ms_gsl/4.0.0-6/include -isystem /sw/slc7_x86-64/libuv/v1.40.0-23/include -isystem /sw/slc7_x86-64/FairRoot/v18.4.9-39/include -isystem /sw/slc7_x86-64/VMC/v2-0-56/include/vmc -isystem /sw/slc7_x86-64/Vc/1.4.1-87/include -isystem /sw/slc7_x86-64/libjalienO2/0.1.4-5/include -fPIC -O2 -std=c++17 -Werror -Wno-error=deprecated-declarations -O2 -g -Wno-unknown-warning-option -fPIC -std=gnu++17 -MD -MT Framework/PhysicsSupport/CMakeFiles/O2lib-FrameworkPhysicsSupport.dir/src/Plugin.cxx.o -MF Framework/PhysicsSupport/CMakeFiles/O2lib-FrameworkPhysicsSupport.dir/src/Plugin.cxx.o.d -o Framework/PhysicsSupport/CMakeFiles/O2lib-FrameworkPhysicsSupport.dir/src/Plugin.cxx.o -c /sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/Framework/PhysicsSupport/src/Plugin.cxx
/sw/SOURCES/O2/build_QualityControl_o2-slc7_x86-64/0/Framework/PhysicsSupport/src/Plugin.cxx:16:10: fatal error: Framework/O2DatabasePDGPlugin.h: No such file or directory
   16 | #include "Framework/O2DatabasePDGPlugin.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

ktf commented:
Weird. Is this in the CI?
